File: bench/compare/postgres/probe_notifications.exs
database_url = System.fetch_env!("DOCKET_BENCH_DATABASE_URL")
uri = URI.parse(database_url)
database = uri.path |> String.trim_leading("/")

credentials =
  case uri.userinfo && String.split(uri.userinfo, ":", parts: 2) do
    nil -> []
    [username] -> [username: URI.decode(username)]
    [username, password] -> [username: URI.decode(username), password: URI.decode(password)]
  end

connection =
  [
    hostname: uri.host,
    port: uri.port || 5432,
    database: database,
    sync_connect: true,
    ssl: [verify: :verify_none]
  ] ++ credentials

channel = "docket_bench_probe"
payload = Base.encode16(:crypto.strong_rand_bytes(12), case: :lower)

{:ok, listener} = Postgrex.Notifications.start_link(connection)
{:ok, reference} = Postgrex.Notifications.listen(listener, channel)
{:ok, sender} = Postgrex.start_link(connection)
_result = Postgrex.query!(sender, "SELECT pg_notify($1, $2)", [channel, payload])

receive do
  {:notification, ^listener, ^reference, ^channel, ^payload} ->
    IO.puts("received=1 payload_matched=true")
after
  5_000 ->
    IO.puts("received=0 payload_matched=false")
    System.halt(1)
end

File: bench/compare/postgres/probe_notifications.py
#!/usr/bin/env python3
"""Probe raw LISTEN/NOTIFY delivery through DOCKET_BENCH_DATABASE_URL."""

from __future__ import annotations

import os
import secrets

import psycopg


def main() -> None:
    database_url = os.environ["DOCKET_BENCH_DATABASE_URL"]
    channel = "docket_bench_probe"
    payload = secrets.token_hex(12)

    with (
        psycopg.connect(database_url, autocommit=True) as listener,
        psycopg.connect(database_url, autocommit=True) as sender,
    ):
        listener.execute(f"LISTEN {channel}")
        sender.execute("SELECT pg_notify(%s, %s)", (channel, payload))
        received = list(listener.notifies(timeout=5.0, stop_after=1))
        listener.execute(f"UNLISTEN {channel}")

    matched = len(received) == 1 and received[0].payload == payload
    print(f"received={len(received)} payload_matched={matched}")
    if not matched:
        raise SystemExit(1)


if __name__ == "__main__":
    main()
